api/server/router/network/network_routes.go

```diff
@@ -82,14 +84,78 @@ func (n *networkRouter) getNetwork(ctx context.Context, w http.ResponseWriter, r
 		return err
 	}
 
-	nw, err := n.backend.FindNetwork(vars["id"])
-	if err != nil {
-		if nr, err := n.cluster.GetNetwork(vars["id"]); err == nil {
-			return httputils.WriteJSON(w, http.StatusOK, nr)
+	term := vars["id"]
+
+	// In case multiple networks have duplicate names, return error.
+	// TODO (yongtang): should we wrap with version here for backward compatibility?
+
+	// First find based on full ID, return immediately once one is found.
+	// If a network appears both in swarm and local, assume it is in local first
+
+	// For full name and partial ID, same the result first, and process later
+	// in case multiple records was found based on the same term
+	listByFullName := map[string]types.NetworkResource{}
+	listByPartialID := map[string]types.NetworkResource{}
+
+	nw := n.backend.GetNetworks()
+	for _, network := range nw {
+		if network.ID() == term {
+			return httputils.WriteJSON(w, http.StatusOK, *n.buildDetailedNetworkResources(network))
+		}
+		if network.Name() == term {
+			// It may not be necessary to check ID collision (not Name collision) here,
+			// if the ID is guaranteed to be unique in local scope
+			if _, ok := listByFullName[network.ID()]; !ok {
+				listByFullName[network.ID()] = *n.buildDetailedNetworkResources(network)
+			}
+		}
+		if strings.HasPrefix(network.ID(), term) {
+			// It may not be necessary to check ID collision (not Name collision) here,
+			// if the ID is guaranteed to be unique in local scope
+			if _, ok := listByPartialID[network.ID()]; !ok {
+				listByPartialID[network.ID()] = *n.buildDetailedNetworkResources(network)
+			}
+		}
+	}
+
+	nr, _ := n.cluster.GetNetworks()
+	for _, network := range nr {
+		if network.ID == term {
+			return httputils.WriteJSON(w, http.StatusOK, network)
+		}
+		if network.Name == term {
+			if _, ok := listByFullName[network.ID]; !ok {
+				listByFullName[network.ID] = network
+			}
+		}
+		if strings.HasPrefix(network.ID, term) {
+			if _, ok := listByPartialID[network.ID]; !ok {
+				listByPartialID[network.ID] = network
+			}
+		}
+	}
+
+	// Find based on full name, returns true only if no duplicates
+	if len(listByFullName) == 1 {
+		for _, v := range listByFullName {
+			return httputils.WriteJSON(w, http.StatusOK, v)
 		}
-		return err
 	}
-	return httputils.WriteJSON(w, http.StatusOK, n.buildDetailedNetworkResources(nw))
+	if len(listByFullName) > 1 {
+		return fmt.Errorf("network %s is ambiguous (%d matches found based on name)", term, len(listByFullName))
+	}
+
+	// Find based on partial ID, returns true only if no duplicates
+	if len(listByPartialID) == 1 {
+		for _, v := range listByPartialID {
```

Comment:
Given you already check the len(list) == 1 condition, I would avoid to write the loop here

```
return httputils.WriteJSON(w, http.StatusOK, listByPartialID[0])
```